src/storage/src/source/reclock/compat.rs

```diff
@@ -102,23 +103,11 @@ where
             .context("error opening persist shard")?;
 
         let upper = write_handle.upper();
-        // We want a leased reader because elsewhere in the code the `as_of`
-        // time may also be determined by another `ReadHandle`, and the pair of
-        // them offer the invariant that we need (that the `as_of` if <= this
-        // `since`). Using a `SinceHandle` here does not offer the same
-        // invariant when paired with a `ReadHandle`.
-        let since = read_handle.since();
+        let since = read_handle.since().clone();
 
         // Allow manually simulating the scenario where the since of the remap
         // shard has advanced too far.
         fail_point!("invalid_remap_as_of");
-        assert!(
```

Comment:
Where are we now doing something that is equivalent to this assert? When/if this doesn't hold, we might produce duplicate updates to the source shard, afaik.

Reply:
I'm curious about this removal too

Reply:
Fair! In this PR we are worried about:

> While the since is normally in advance of the uppers, this does not appear to be the case when the source is being created or shutting down.

If we want to choose an as-of based on the uppers of the shard and not the since of the shard, the condition this assert checks for will become normal. So this PR makes some effort to make sure it is also okay. For example, in the startup case the global since of the shard is way ahead of the upper, and thus the as-of... but we happen to know that no data will be produced in that interval, so reading from the since and accumulating to the as-of will produce the same result.

(Ideally the since would be held not past the upper in the controller so the shard was always readable, but that's a separate can of worms...)

Reply:
I pushed a commit that re-adds the assert with a carveout for `[0]`. Would that sort of thing be better?